### PR TITLE
Feat: support to manage the configs by the CLI, UI, and workflow.

### PR DESCRIPTION
**Pull request by barnettZQG**
Signed-off-by: barnettZQG <barnett.zqg@gmail.com>

### Description of your changes

This PR refactors the config management. Specific changes include:

#### 1. Added the CUE template package.

There are many features depending on the CUE template. Such as the Config, Addon, Definitions, etc. To support the config render directly, I write a package `pkg/cue/template` to provide the render, validation, and generate the schema features based on CUE. The Addon and Definition depend on this package to generate the API schema.

#### 3. Change the config render from the Application to the CUE template.

Depending on the Application to render the config is too complex and unsafe. The concept is unclear. For the config content, we only need a simple template or do not need the template. Now, use the Secret as the entity of the config. Users could define a template to tell KubeVela how to generate the Secret.  By default, KubeVela uses the template name as the type of the Secret.

#### 4. The distribution of the config could be managed by users.

The distribution of the config still depends on the Application. But KubeVela does not create the application auto, users could distribute the config by CLI and UI. They could find the distribution status in the config list.

#### 5. The config could be updated. 

For the not sensitive configs, which could be edited and distributed again. 

#### 6. Users could manage the config by CLI, API, UI, and Workflow.

* CLI 

Manage the templates:

```bash

# Create or update a template, this template will be saved as a Configmap.
vela config-template apply --file(-f)  [--namespace(-n)]

vela config-template list [-n] [-A]

vela config-template delete <template-name> [-n]

# Show the document of the template properties.
vela config-template show <template-name> [-n]
```

Manage the configs:

```bash

# Create a config from a file or the command args. If specify the target, this config will be distributed.
vela config create xxx --template(-t) [Properties KV] [--target] [--namespace(-n)] [-f]

# Distribute the config to the target namespace. 
vela config distribute <config-name> --target(-t) <[cluster/]namespace> 

# Recall the config from all targets.
vela config distribute <config-name> --recall

vela config list [-n] [-A] [--template(-t)]

vela config delete <config-name> [--namespace(-n)]
```

* UI

image

*The platform config management*

image

*The project config management*

#### 7. Support to write the configs to the Nacos server.

If users want to write the config to the Nacos server, this feature is very suitable. If the template includes the Nacos definition, the config content could be sent to the Nacos server. Here is a example template:

```cue
metadata: {
	name:        "nacos-config"
	alias:       "Nacos Configuration"
	description: "Write the configuration to the nacos"
	sensitive:   false
	scope:       "system"
}

template: {
	nacos: {
		// The endpoint can not references the parameter.
		endpoint: {
			name:      "nacos"
			namespace: "vela-system"
		}
		format: parameter.contentType

		// could references the parameter
		metadata: {
			dataId: parameter.dataId
			group:  parameter.group
			if parameter.appName != _|_ {
				appName: parameter.appName
			}
			if parameter.namespaceId != _|_ {
				namespaceId: parameter.namespaceId
			}
			if parameter.tenant != _|_ {
				tenant: parameter.tenant
			}
			if parameter.tag != _|_ {
				tenant: parameter.tag
			}
		}
		content: parameter.content
	}
	parameter: {
		// +usage=Configuration ID
		dataId: string
		// +usage=Configuration group
		group: *"DEFAULT_GROUP" | string
		// +usage=The configuration content.
		content: {
			...
		}
		contentType: *"json" | "yaml" | "properties" | "toml"
		// +usage=The app name of the configuration
		appName?: string
		// +usage=The namespaceId of the configuration
		namespaceId?: string
		// +usage=The tenant, corresponding to the namespace ID field of Nacos
		tenant?: string
		// +usage=The tag of the configuration
		tag?: string
	}
}
```

BTW, the config management could as a solution for #4605. 

## Deprecated

* The vela provider commands. The capabilities of these commands are provided by config-related commands.

```bash
vela provider add <provider-type>  => vela config create --template <provider-type>

vela provider delete <provider-name> =>  vela config delete <provider-name>
```

* Deprecated the PackageDiscover package.

## API Changes

The following APIs have breaking changes:

* Path: `/api/v1/config_types/*` Deprecated

* Path: `/api/v1/projects/<name>/configs` Response body changed.

Reference: https://github.com/kubevela/kubevela/pull/4794/files#diff-1a8a07dce160f7a1e106dc6e0881d8e7612ca4d8cd04011622a32f7b304d5483

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: apis/types/types.go [hunk 1]
```diff
@@ -156,11 +164,15 @@ const (
...
 	TerraformProvider = "terraform-provider"
 	// DexConnector is the config type for dex connector
-	DexConnector = "config-dex-connector"
+	DexConnector = "dex-connector"
 	// ImageRegistry is the config type for image registry
-	ImageRegistry = "config-image-registry"
+	ImageRegistry = "image-registry"
 	// HelmRepository is the config type for Helm chart repository
-	HelmRepository = "config-helm-repository"
+	HelmRepository = "helm-repository"
```
Comment (wonderflow) [inline apis/types/types.go:171]: will these changes cause incompatibility?

Comment (barnettZQG) [reply apis/types/types.go:171]: The old secrets include the `helm-repository` lable, so it is same with new.

File: charts/vela-core/templates/defwithtemplate/config-image-registry.yaml [hunk 2]
```diff
@@ -1,86 +0,0 @@
-# Code generated by KubeVela templates. DO NOT EDIT. Please edit the original cue file.
-# Definition source cue file: vela-templates/definitions/internal/config-image-registry.cue
```
Comment (wonderflow) [inline charts/vela-core/templates/defwithtemplate/config-image-registry.yaml:2]: why this file deleted?

File: docs/examples/config/image-registry/app-validate-imagePullSecret.yaml [hunk 3]
```diff
@@ -1,14 +0,0 @@
-apiVersion: core.oam.dev/v1beta1
-kind: Application
```
Comment (wonderflow) [inline docs/examples/config/image-registry/app-validate-imagePullSecret.yaml:2]: you need to write a new example of config

File: pkg/addon/addon.go [hunk 4]
```diff
@@ -318,7 +324,7 @@ func GetUIDataFromReader(r AsyncReader, meta *SourceMeta, opt ListOptions) (*UID
 		}
 		err := genAddonAPISchema(addon)
 		if err != nil {
-			return nil, fmt.Errorf("fail to generate openAPIschema for addon %s : %w", meta.Name, err)
+			return nil, fmt.Errorf("fail to generate openAPIschema for addon %s : %w (parameters: %s)", meta.Name, err, addon.Parameters)
```
Comment (wonderflow) [inline pkg/addon/addon.go:327]:
```suggestion
			return nil, fmt.Errorf("fail to generate openAPIschema for addon %s : %w (parameter: %s)", meta.Name, err, addon.Parameters)
```

We use `parameter` for cue paramter overall.

File: pkg/addon/addon.go [hunk 5]
```diff
@@ -454,6 +461,22 @@ func readDefFile(a *UIData, reader AsyncReader, readPath string) error {
...
+func readConfigTemplateFile(a *UIData, reader AsyncReader, readPath string) error {
+	b, err := reader.ReadFile(readPath)
+	if err != nil {
+		return err
+	}
+	filename := path.Base(readPath)
+	if filepath.Ext(filename) != ".cue" {
+		return nil
+	}
+	fmt.Println("read the template file " + filename)
```
Comment (wonderflow) [inline pkg/addon/addon.go:474]: debug message?

File: pkg/apiserver/domain/service/config.go [hunk 6]
```diff
@@ -64,425 +54,271 @@ func NewConfigService() ConfigService {
...
-				}
-				mergedTargets = append(mergedTargets, ApplicationDeployTarget{Namespace: c.Namespace,
-					Clusters: interfaceToStringSlice(clusters.ToSlice())})
-			}
-		}
-		if !hasSameNamespace {
-			mergedTargets = append(mergedTargets, c)
-		}
+// ApplyConfigDistribution distribute the configs to the target namespaces.
+func (u *configServiceImpl) ApplyConfigDistribution(ctx context.Context, project string, req apis.ApplyConfigDistributionRequest) error {
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/config.go:236]:
```suggestion
func (u *configServiceImpl) DistributeConfigs(ctx context.Context, project string, req apis.ApplyConfigDistributionRequest) error {
```

File: pkg/apiserver/utils/bcode/016_config.go [hunk 7]
```diff
@@ -14,18 +14,12 @@ See the License for the specific language governing permissions and
...
 
-import (
-	v1 "k8s.io/api/core/v1"
+var (
+	// ErrSensitiveConfig means the config can not be read
+	ErrSensitiveConfig = NewBcode(400, 16001, "the config is sensitive")
 
-	"github.com/oam-dev/kubevela/apis/types"
+	// ErrNoConfigOrTarget means there is no target or config when creating the distribution.
+	ErrNoConfigOrTarget = NewBcode(400, 16002, "the config list or the target list can not be empty")
```
Comment (wonderflow) [inline pkg/apiserver/utils/bcode/016_config.go:24]:
```suggestion
	ErrNoConfigOrTarget = NewBcode(400, 16002, "you must specify the config name and destination to distribute")
```

File: pkg/config/factory.go [hunk 8]
```diff
@@ -0,0 +1,862 @@
...
+const SaveObjectReference = "objects-reference"
+
+// TemplateConfigMapNamePrefix the prefix of the configmap name
+const TemplateConfigMapNamePrefix = "config-template-"
+
+// ErrSensitiveConfig means this config can not be read directly.
+var ErrSensitiveConfig = fmt.Errorf("the config is sensitive")
+
+// ErrNoConfigOrTarget means the config or the target is empty
+var ErrNoConfigOrTarget = fmt.Errorf("the config or the target is empty")
```
Comment (wonderflow) [inline pkg/config/factory.go:66]:
```suggestion
var ErrNoConfigOrTarget = errors.New(""you must specify the config name and destination to distribute")
```

File: pkg/config/factory.go [hunk 9]
```diff
@@ -0,0 +1,862 @@
...
+const SaveInputPropertiesKey = "input-properties"
+
+// SaveObjectReference define the key name for saving the outputs objects reference metadata in the secret
+const SaveObjectReference = "objects-reference"
+
+// TemplateConfigMapNamePrefix the prefix of the configmap name
+const TemplateConfigMapNamePrefix = "config-template-"
+
+// ErrSensitiveConfig means this config can not be read directly.
+var ErrSensitiveConfig = fmt.Errorf("the config is sensitive")
```
Comment (wonderflow) [inline pkg/config/factory.go:63]: Please use `errors.New`  instead of `fmt.Errorf` for prue string error

File: pkg/config/factory.go [hunk 10]
```diff
@@ -0,0 +1,862 @@
...
+
+// ErrNoConfigOrTarget means the config or the target is empty
+var ErrNoConfigOrTarget = fmt.Errorf("the config or the target is empty")
+
+// ErrNotFoundDistribution means the app of the distribution is not exist.
+var ErrNotFoundDistribution = fmt.Errorf("the distribution is not found")
```
Comment (wonderflow) [inline pkg/config/factory.go:69]: Please use `errors.New`  instead of `fmt.Errorf` for prue string error

File: pkg/config/factory.go [hunk 11]
```diff
@@ -0,0 +1,862 @@
...
+type NamespacedName struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+// Template This is the spec of the config template, parse from the cue script.
+type Template struct {
+	NamespacedName
+	Alias       string `json:"alias"`
+	Description string `json:"description"`
```
Comment (wonderflow) [inline pkg/config/factory.go:81]: omitempty?

File: pkg/config/factory.go [hunk 12]
```diff
@@ -0,0 +1,862 @@
...
+			configs = append(configs, it)
+		}
+	}
+	return configs, nil
+}
+
+func (k *kubeConfigFactory) DeleteConfig(ctx context.Context, namespace, name string) error {
+	var secret v1.Secret
+	c, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
```
Comment (wonderflow) [inline pkg/config/factory.go:599]: can we use a global timeout context for this?

File: charts/vela-core/templates/defwithtemplate/config-image-registry.yaml (hunk 2 above)
Comment (barnettZQG) [reply charts/vela-core/templates/defwithtemplate/config-image-registry.yaml:2]: Move to the VelaUX addon.

File: docs/examples/config/image-registry/app-validate-imagePullSecret.yaml (hunk 3 above)
Comment (barnettZQG) [reply docs/examples/config/image-registry/app-validate-imagePullSecret.yaml:2]: Added

**Review by wonderflow — CHANGES_REQUESTED** [review]
please add tests for new APIs

File: references/cli/config.go [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+	Targets   []string
+	Config    string
+	Namespace string
+	Recalled  bool
+}
+
+// ConfigApplyCommandOptions the options of the command that apply the config.
+type ConfigApplyCommandOptions struct {
+	Template   string
+	Namespace  string
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply references/cli/config.go:260]: The config could save in any namespace. There is a policy in VelaUX that the configs in the vela-system namespace could be shared.

File: references/cli/config.go [hunk 14]
```diff
@@ -0,0 +1,625 @@
...
+			if err := options.parseProperties(args[1:]); err != nil {
+				return err
+			}
+			configItem, err := inf.ParseConfig(context.Background(), config.NamespacedName{
+				Name:      name,
+				Namespace: namespace,
+			}, config.Metadata{
+				NamespacedName: config.NamespacedName{
+					Name:      options.Name,
+					Namespace: options.Namespace,
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply references/cli/config.go:419]: The default value of the namespace is "vela-system".

File: pkg/config/factory.go [hunk 15]
```diff
@@ -0,0 +1,881 @@
...
+	return item, nil
+}
+
+// CreateOrUpdateConfig create or update the config.
+// Write the expand config to the target server.
+func (k *kubeConfigFactory) CreateOrUpdateConfig(ctx context.Context, i *Config, ns string) error {
+	var secret v1.Secret
+	if err := k.cli.Get(ctx, pkgtypes.NamespacedName{Namespace: i.Namespace, Name: i.Name}, &secret); err == nil {
+		if secret.Labels[types.LabelConfigType] != i.Template.Name {
+			return ErrConfigExist
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply pkg/config/factory.go:543]: The template name can not be changed when updating the config. Here only check the template name.

File: references/cli/config.go [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+			}
+			if err := options.parseProperties(args[1:]); err != nil {
+				return err
+			}
+			configItem, err := inf.ParseConfig(context.Background(), config.NamespacedName{
+				Name:      name,
+				Namespace: namespace,
+			}, config.Metadata{
+				NamespacedName: config.NamespacedName{
+					Name:      options.Name,
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply references/cli/config.go:418]: The first name is the template name, and here is the config name.

File: references/cli/config.go [hunk 17]
```diff
@@ -0,0 +1,625 @@
...
+					Writer: streams.Out,
+					Reader: bufio.NewReader(streams.In),
+				}
+				if !assumeYes {
+					userConfirmation := userInput.AskBool("Do you want to recall this config", &UserInputOptions{assumeYes})
+					if !userConfirmation {
+						return fmt.Errorf("stopping recalling")
+					}
+				}
+				if err := inf.DeleteDistribution(context.Background(), options.Namespace, name); err != nil {
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply references/cli/config.go:534]: `--recall` means recall the config from all targets. Users could first use the `vela config distribute xxx -t <cluster/namespace>` command to distribute the config to the multi-clusters.

File: references/cli/config.go [hunk 18]
```diff
@@ -0,0 +1,625 @@
...
+// DistributeConfigCommandOptions the options of the command that distribute the config.
+type DistributeConfigCommandOptions struct {
+	Targets   []string
+	Config    string
+	Namespace string
+	Recalled  bool
+}
+
+// ConfigApplyCommandOptions the options of the command that apply the config.
+type ConfigApplyCommandOptions struct {
```
[the remark this answers is not in the document]
Comment (barnettZQG) [reply references/cli/config.go:258]: I will change it.

File: apis/types/types.go [hunk 19]
```diff
@@ -74,7 +74,7 @@ const (
 	AnnoIngressControllerHTTPPort = "ingress.controller/http-port"
 	// AnnoIngressControllerHost define ingress controller externally host
 	AnnoIngressControllerHost = "ingress.controller/host"
-	// LabelConfigType is the label for config type
+	// LabelConfigType is the label for the config type
```
Comment (wonderflow) [inline apis/types/types.go:77]:
```suggestion
	// LabelConfigType is the label marked on all resources generated from config type 
```

File: pkg/apiserver/domain/service/config.go [hunk 20]
```diff
@@ -64,425 +54,299 @@ func NewConfigService() ConfigService {
...
+			Alias:       template.Alias,
+			Name:        template.Name,
+			Description: template.Description,
+			Namespace:   template.Namespace,
+			Scope:       template.Scope,
+			Sensitive:   template.Sensitive,
+			CreateTime:  template.CreateTime,
+		},
+		APISchema: template.Schema,
+		// TODO: Support to define the custom UI schema in the template cue script.
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/config.go:122]: does it align with the uischema of definitions?

File: pkg/apiserver/domain/service/image.go [hunk 21]
```diff
@@ -93,51 +94,43 @@ func (i *imageImpl) GetImageInfo(ctx context.Context, project, secretName, image
...
-			types.LabelConfigIdentifier: registryDomain,
-		}); err != nil {
-		log.Logger.Warnf("fail to list the docker registries, %s", err.Error())
-	}
-	var selectSecret []*corev1.Secret
-	var selectSecretNames []string
+
+	registries, err := i.ListImageRepos(ctx, project)
+	if err != nil {
+		imageInfo.Message = "There is no registry."
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/image.go:100]: you should log the real error for debug instead of erase it.

File: pkg/config/factory.go [hunk 22]
```diff
@@ -0,0 +1,881 @@
...
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
+	}
+	sensitive, err := value.GetBool("metadata", "sensitive")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
+	}
+	templateValue, err := value.LookupValue("template")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
```
Comment (wonderflow) [inline pkg/config/factory.go:238]: why don't return error above?

File: pkg/config/factory.go [hunk 23]
```diff
@@ -0,0 +1,881 @@
...
+	if defaultName != "" {
+		name = defaultName
+	}
+	schema, err := cueScript.ParsePropertiesToSchema()
+	if err != nil {
+		return nil, fmt.Errorf("the properties of the cue script is invalid:%w", err)
+	}
+	alias, err := value.GetString("metadata", "alias")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the alias from the template metadata: %s", err.Error())
```
Comment (wonderflow) [inline pkg/config/factory.go:226]: alias not necessary, no need to report error

File: pkg/config/factory.go [hunk 24]
```diff
@@ -0,0 +1,881 @@
...
+	}
+	scope, err := value.GetString("metadata", "scope")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
```
Comment (wonderflow) [inline pkg/config/factory.go:230]: don't we have default data? we shouldn't report warning message in this case.

File: pkg/config/factory.go [hunk 25]
```diff
@@ -0,0 +1,881 @@
...
+	}
+	sensitive, err := value.GetBool("metadata", "sensitive")
+	if err != nil && !IsFieldNotExist(err) {
+		klog.Warningf("fail to get the scope from the template metadata: %s", err.Error())
```
Comment (wonderflow) [inline pkg/config/factory.go:234]: does it necessary data?

File: pkg/config/factory.go [hunk 26]
```diff
@@ -0,0 +1,881 @@
...
+func (k *kubeConfigFactory) CreateOrUpdateConfigTemplate(ctx context.Context, ns string, it *Template) error {
+	if ns != "" {
+		it.ConfigMap.Namespace = ns
+	}
+	return k.apiApply.Apply(ctx, it.ConfigMap, apply.DisableUpdateAnnotation(), apply.Quiet())
+}
+
+func convertConfigMap2Template(cm v1.ConfigMap) (*Template, error) {
+	if cm.Labels == nil || cm.Annotations == nil {
+		return nil, fmt.Errorf("this configmap is not a valid template")
```
Comment (wonderflow) [inline pkg/config/factory.go:299]:
```suggestion
		return nil, fmt.Errorf("this configmap is not a valid config-template")
```

File: pkg/config/factory.go [hunk 27]
```diff
@@ -0,0 +1,881 @@
...
+	}
+	if cm.Data["schema"] != "" {
+		var schema openapi3.Schema
+		err := yaml.Unmarshal([]byte(cm.Data["schema"]), &schema)
+		if err != nil {
+			return nil, fmt.Errorf("fail to parse the schema: %w", err)
+		}
+		it.Schema = &schema
+	}
+	if cm.Data["expanded-writer"] != "" {
```
Comment (wonderflow) [inline pkg/config/factory.go:321]: let's use a concrete string for it

File: references/cli/config.go [hunk 28]
```diff
@@ -0,0 +1,631 @@
...
+			}
+			options.Name = args[0]
+			userInput := &UserInput{
+				Writer: streams.Out,
+				Reader: bufio.NewReader(streams.In),
+			}
+			if !assumeYes {
+				userConfirmation := userInput.AskBool("Do you want to delete this template", &UserInputOptions{assumeYes})
+				if !userConfirmation {
+					return fmt.Errorf("stopping deleting")
```
Comment (wonderflow) [inline references/cli/config.go:234]:
```suggestion
					return fmt.Errorf("deleting stopped")
```

File: pkg/config/factory.go (hunk 25 above)
Comment (barnettZQG) [reply pkg/config/factory.go:234]: The alias, scope, and sensitive are not required fields. So here ignore the not exist error, and print the other errors.

**Review by wonderflow — APPROVED** [review]

**Review by FogDong — APPROVED** [review]

**Review by Somefive — APPROVED** [review]

**Pull request merged**